client-tez/src/main/java/org/apache/tez/runtime/library/common/shuffle/orderedgrouped/RssInMemoryMerger.java

```diff
@@ -0,0 +1,221 @@
+  public void merge(List<MapOutput> inputs) throws IOException, InterruptedException {
+    if (inputs == null || inputs.size() == 0) {
+      return;
+    }
+    numMemToRemoteMerges.increment(1);
+    inputContext.notifyProgress();
+
+    this.srcTaskIdentifier = inputs.get(0).getAttemptIdentifier();
+    List<TezMerger.Segment> inMemorySegments = new ArrayList();
+
+    manager.createInMemorySegments(inputs, inMemorySegments, 0);
+    int noInMemorySegments = inMemorySegments.size();
+
+    String mergedFile =
+        String.format(
+            SPILL_FILE_PATTERN,
+            appAttemptId,
+            inputContext.getUniqueIdentifier(),
+            srcTaskIdentifier.getInputIdentifier(),
+            srcTaskIdentifier.getSpillEventId());
+    this.outputPath = new Path(spillPath, mergedFile);
+
+    Writer writer = null;
+    long outFileLen = 0;
+    try {
+      writer =
+          new Writer(
+              conf,
+              remoteFs,
+              outputPath,
+              ConfigUtils.getIntermediateInputKeyClass(conf),
+              ConfigUtils.getIntermediateInputValueClass(conf),
+              codec,
+              null,
+              null);
+
+      TezRawKeyValueIterator rIter;
+      LOG.info("Initiating in-memory merge with " + noInMemorySegments + " segments...");
+
+      // When factor is greater or equal to the size of segements, we will ignore
+      // intermediate merger, so tmpeDir and tmpFs is useless
+      Path tmpDir = null;
+      FileSystem tmpFs = null;
+      rIter =
+          TezMerger.merge(
+              conf,
+              tmpFs,
+              ConfigUtils.getIntermediateInputKeyClass(conf),
+              ConfigUtils.getIntermediateInputValueClass(conf),
+              inMemorySegments,
+              inMemorySegments.size(),
+              tmpDir,
+              ConfigUtils.getIntermediateInputKeyComparator(conf),
+              manager.getProgressable(),
+              spilledRecordsCounter,
+              null,
+              additionalBytesRead,
+              null);
+
+      if (null == combiner) {
+        TezMerger.writeFile(
+            rIter,
+            writer,
+            manager.getProgressable(),
+            TezRuntimeConfiguration.TEZ_RUNTIME_RECORDS_BEFORE_PROGRESS_DEFAULT);
+      } else {
+        runCombineProcessor(rIter, writer);
+      }
+      writer.close();
+      additionalBytesWritten.increment(writer.getCompressedLength());
+      writer = null;
+      outFileLen = remoteFs.getFileStatus(outputPath).getLen();
+      LOG.info(
+          inputContext.getUniqueIdentifier()
+              + " Merge of the "
+              + noInMemorySegments
+              + " files in-memory complete."
+              + " Remote file is "
+              + outputPath
+              + " of size "
+              + outFileLen);
+    } catch (IOException e) {
+      remoteFs.delete(outputPath, true);
+      throw e;
+    } finally {
+      if (writer != null) {
+        writer.close();
```

Comment:
Why do we close this writer in the Line#179?

Reply:
> Why do we close this writer in the Line#179?

@jerqi  
Both TezMerger.merge and TezMerger.writeFile may throw exception. If throw exception, writer will not be closed.
By the way, both Line#179 and Line#181 may be useless, I will remove it.

Reply:
Could you add some comments to explain why? It's a little weird.

Reply:
> Could you add some comments to explain why? It's a little weird.

OK, I have comment it.